worker/streak.js

```diff
@@ -2,59 +2,56 @@ import { notifyNewStreak, notifyStreakLost } from '@/lib/webPush'
-      -- a bunch of mutations
-      streak_insert AS (
-        INSERT INTO "Streak" ("userId", "startedAt", "type", created_at, updated_at)
-        SELECT id, (now() AT TIME ZONE 'America/Chicago' - interval '1 day')::date, ${type}::"StreakType", now_utc(), now_utc()
-        FROM new_streaks
-      ), user_update_new_streaks AS (
-        UPDATE users SET ${getStreakColumn(type)} = 1 FROM new_streaks WHERE new_streaks.id = users.id
-      ), user_update_end_streaks AS (
-        UPDATE users SET ${getStreakColumn(type)} = NULL FROM ending_streaks WHERE ending_streaks.id = users.id
-      ), user_update_extend_streaks AS (
-        UPDATE users
-        SET ${getStreakColumn(type)} = (now() AT TIME ZONE 'America/Chicago')::date - extending_streaks.started_at
-        FROM extending_streaks WHERE extending_streaks.id = users.id
-      )
-      UPDATE "Streak"
-      SET "endedAt" = (now() AT TIME ZONE 'America/Chicago' - interval '1 day')::date, updated_at = now_utc()
-      FROM ending_streaks
-      WHERE ending_streaks.id = "Streak"."userId" AND "endedAt" IS NULL AND "type" = ${type}::"StreakType"
-      RETURNING "Streak".*`
+  const type = 'COWBOY_HAT'
+  const endingStreaks = await models.$queryRaw`
+    WITH day_streaks (id) AS (
+      ${getStreakQuery(type)}
+    ), existing_streaks (id, started_at) AS (
+      SELECT "userId", "startedAt"
+      FROM "Streak"
+      WHERE "Streak"."endedAt" IS NULL
+      AND "type" = ${type}::"StreakType"
+    ), new_streaks (id) AS (
+      SELECT day_streaks.id
+      FROM day_streaks
+      LEFT JOIN existing_streaks ON existing_streaks.id = day_streaks.id
+      WHERE existing_streaks.id IS NULL
+    ), ending_streaks (id) AS (
+      SELECT existing_streaks.id
+      FROM existing_streaks
+      LEFT JOIN day_streaks ON existing_streaks.id = day_streaks.id
+      WHERE day_streaks.id IS NULL
+    ), extending_streaks (id, started_at) AS (
+      SELECT existing_streaks.id, existing_streaks.started_at
+      FROM existing_streaks
+      JOIN day_streaks ON existing_streaks.id = day_streaks.id
+    ),
+    -- a bunch of mutations
+    streak_insert AS (
+      INSERT INTO "Streak" ("userId", "startedAt", "type", created_at, updated_at)
+      SELECT id, (now() AT TIME ZONE 'America/Chicago' - interval '1 day')::date, ${type}::"StreakType", now_utc(), now_utc()
+      FROM new_streaks
+    ), user_update_new_streaks AS (
+      UPDATE users SET "streak" = 1 FROM new_streaks WHERE new_streaks.id = users.id
+    ), user_update_end_streaks AS (
+      UPDATE users SET "streak" = NULL FROM ending_streaks WHERE ending_streaks.id = users.id
+    ), user_update_extend_streaks AS (
+      UPDATE users
+      SET "streak" = (now() AT TIME ZONE 'America/Chicago')::date - extending_streaks.started_at::date
+      FROM extending_streaks WHERE extending_streaks.id = users.id
+    )
+    UPDATE "Streak"
+    SET "endedAt" = (now() AT TIME ZONE 'America/Chicago' - interval '1 day')::date, updated_at = now_utc()
```

Comment:
@huumn why is this using `- interval '1 day'`?

Shouldn't we start to count from 1 and not from 0 for streaks? Like if you got a cowboy hat and then lose it, you at least kept it for one day since we run this at midnight (= next day)?

Asking because I ran this on the same day with `COWBOY_HAT_THRESHOLD` set to 1000 to make me lose my hat and then it shows up as -1 days. I expected it to say 0 days.

Reply:
That statement in total means "yesterday's date" because this is run shortly after midnight.